sideboard/internal/connection_checker.py

```diff
@@ -1,6 +1,7 @@
 from __future__ import unicode_literals, print_function
```

Comment:
MAGFest doesn't use or care about any of this, but let me go on a bit of rant here.  Sometimes the hardest thing to get working in a distributed system is just getting a thing to talk to another thing.  Like, let's say that you find that for whatever reason Server A isn't successfully making API calls to Server B.  Here's a list of common reasons I've seen:

- The config file has the URL in the wrong format.  For example, people just put foo.com instead of https://foo.com
The other server exists but DNS won't resolve its hostname.  (This happens more often than you'd think, since a lot of intranets have asymmetric DNS rules where Server A can resolve Server B's hostname but not the other way around.)
- We can't establish a socket connection in the first place.  Here are the most common reasons in my experience:
- Client is trying to connect to the wrong port number.
- Server is listening on the wrong port number.
- Iptables rules on the client are blocking outbound traffic.
- Iptables rules on the server are blocking inbound traffic
- Some firewall between the two servers is blocking the traffic.
- We can open a socket, but the SSL handshake fails.  What follows are the most common causes of that:
- Client isn't using raw HTTP but the server is on HTTPS.
- Client tries to perform an SSL handshake when the server is doing plain HTTP.
- Server fails to validate the client.  Most common causes in my experience are as follows:
- Client didn't provide a client cert.
- Client provided a client cert that was signed by the wrong CA.
- Client provided a client cert signed by the correct CA but the CN was missing or incorrect.
- Client has a valid client cert, but we specified the wrong path.  (Many Python module just ignore an invalid path to a cert, so you get a "SSL handshake failed" error instead of a "invalid path to file" error.)
- Client fails to validate the server.  Here are the most common causes of that:
- Client was checking the wrong CA.
- Client has the correct CA but was specified the wrong path to the CA file (see above).
- Hostname we connected to doesn't match the server cert's CN.
- SSL version or library incompatibility between client and server.

I've seen all of the above happen multiple times in multiple different situations.

Sideboard therefore offers some utilities for checking all of the API endpoints which it knows about, but iterating through all of them and performing the following steps:
- can we resolve DNS for it
- can we open a regular socket connection to it
- can we perform an SSL handskake on the connection
- can we validate the server's cert
- can we actually make a ``sideboard.poll`` API call

That at least gives us some idea of what's failing so that it narrows down where to look.

MAGFest may start caring about this if/when we start splitting out functionality out across multiple servers.  For example, we may eventually have an ``uber_emails`` plugin which communicates with the main Uber servers via an API but runs separately so as to not make us worry about how our email code affects performance.